mlir/lib/Analysis/CallGraph.cpp

```diff
@@ -93,42 +127,41 @@ static void computeCallGraph(Operation *op, CallGraph &cg,
 
   for (Region &region : op->getRegions())
     for (Operation &nested : region.getOps())
-      computeCallGraph(&nested, cg, symbolTable, parentNode, resolveCalls);
+      computeCallGraph(&nested, cg, symbolTable, parentNode, resolveSymbolRef);
 }
 
 CallGraph::CallGraph(Operation *op)
     : externalCallerNode(/*callableRegion=*/nullptr),
       unknownCalleeNode(/*callableRegion=*/nullptr) {
   // Make two passes over the graph, one to compute the callables and one to
-  // resolve the calls. We split these up as we may have nested callable objects
-  // that need to be reserved before the calls.
+  // resolve the calls and symbol references. We split these up as we may have
+  // nested callable objects that need to be reserved before the calls.
   SymbolTableCollection symbolTable;
-  computeCallGraph(op, *this, symbolTable, /*parentNode=*/nullptr,
-                   /*resolveCalls=*/false);
-  computeCallGraph(op, *this, symbolTable, /*parentNode=*/nullptr,
-                   /*resolveCalls=*/true);
+  computeCallGraph(op, *this, symbolTable, getEntryNode(),
+                   /*resolveSymbolRef=*/false);
+  computeCallGraph(op, *this, symbolTable, getEntryNode(),
+                   /*resolveSymbolRef=*/true);
 }
 
 /// Get or add a call graph node for the given region.
 CallGraphNode *CallGraph::getOrAddNode(Region *region,
                                        CallGraphNode *parentNode) {
-  assert(region && isa<CallableOpInterface>(region->getParentOp()) &&
+  Operation *parentOp = region->getParentOp();
+  assert(region && isa<CallableOpInterface>(parentOp) &&
          "expected parent operation to be callable");
   std::unique_ptr<CallGraphNode> &node = nodes[region];
   if (!node) {
     node.reset(new CallGraphNode(region));
 
     // Add this node to the given parent node if necessary.
-    if (parentNode) {
+    assert(parentNode && "expected non-empty parent node");
+    if (!parentNode->isExternal()) {
       parentNode->addChildEdge(node.get());
-    } else {
-      // Otherwise, connect all callable nodes to the external node, this allows
-      // for conservatively including all callable nodes within the graph.
-      // FIXME This isn't correct, this is only necessary for callable nodes
-      // that *could* be called from external sources. This requires extending
-      // the interface for callables to check if they may be referenced
-      // externally.
-      externalCallerNode.addAbstractEdge(node.get());
+    } else if (auto symbolOp = dyn_cast<SymbolOpInterface>(parentOp)) {
+      // Otherwise, connect all symbol nodes with public visibility
+      // to the entry node, which may be referenced externally.
+      if (!symbolOp.canDiscardOnUseEmpty())
+        parentNode->addRefEdge(node.get());
```

Comment:
What's the motivation for checking for discardability rather than visibility?

Reply:
Although the default implementation of `canDiscardOnUseEmpty` is `getVisibility() != Visibility::Public`, I think it's a more accurate condition to indicate whether this node may be referenced somewhere. e.g. A public symbol with  `canDiscardOnUseEmpty == true` must not be referenced by external node, but a private symbol with `canDiscardOnUseEmpty == false` should be considered as having potential references.

Reply:
Hmm, yeah I think that makes sense to me for now. Could you put this explanation in a comment?

Also what about the case the `parentOp` is not a symbol op? Should we also then add a ref edge conservatively as was previously the case?

Reply:
After careful consideration, I use `symbolOp.isPublic() || !symbolOp.canDiscardOnUseEmpty()` as a conservatice condition. A public symbol with `canDiscardOnUseEmpty == true` is more like a bug, rather than a possible situation.

Oh, right, it's necessary to add a ref edge for non-symbolic callable node if no parent node. But I'm not sure it's meaningful to define a anonymous callable within non-callable operation.